src/llmq/instantsend.cpp

```diff
@@ -751,56 +751,57 @@ void CInstantSendManager::HandleNewInstantSendLockRecoveredSig(const llmq::CReco
     pendingInstantSendLocks.emplace(hash, std::make_pair(-1, islock));
 }
 
-void CInstantSendManager::ProcessMessage(const CNode& pfrom, const std::string& msg_type, CDataStream& vRecv)
+PeerMsgRet CInstantSendManager::ProcessMessage(const CNode& pfrom, gsl::not_null<PeerManager*> peerman, const std::string& msg_type, CDataStream& vRecv)
 {
-    if (!IsInstantSendEnabled()) {
-        return;
-    }
+    if (IsInstantSendEnabled() && msg_type == NetMsgType::ISDLOCK) {
+        if (m_peerman == nullptr) {
+            m_peerman = peerman;
+        }
+        // we should never use one CInstantSendManager with different PeerManager
+        assert(m_peerman == peerman);
```

Comment:
this scares me; why can this not be initialized in the CInstantSendManager constructor? 

This feels like it introduces a data race

Reply:
There's is already race but in constructor.

https://github.com/dashpay/dash/pull/5792/files#diff-3ce892d5580c8044e0b4ebc4d1fe596d919788682fc38ccd66913508ff8fc43dR22

Due to that LLMQContext accept an unique_ptr as an argument because PeerManager would be initialized after initializing of these classes (but required by them):
 - CDKGSessionManager (eliminated by this PR)
 - CSigningManager (eliminated by this PR)
 - CSigSharesManager (still here)
 - CInstantSendManager (eliminated by this PR)

When `ProcessMessage` is called first time, PeerManager is initialized; before any call of Process Message -> there's no any messages -> m_peerman is not needed because we can't ban anyone if no one send messages yet.

Reply:
Well; there's definitely still a race here if two threads call this ProcessMessage function while m_peerman is null

Reply:
there's no actually race because network messages are accepted in one thread. 

But it may change in future. Here's a fix for it: https://github.com/dashpay/dash/pull/5792/commits/7286643443d58ba500c090f6eec1aa44f73569a4